File: app/models/submission.py
"""A blank submission is not student evidence and must never reach grading.

Otherwise an accidental submit is recorded as a failure, mutating mastery and
generating downstream work for code that does not exist.
"""

EMPTY_SUBMISSION_MESSAGE = "Enter some code before submitting."


def is_blank_submission(code: str | None) -> bool:
    return code is None or code.strip() == ""

File: web/lib/submission.ts
export const EMPTY_SUBMISSION_MESSAGE = "Enter some code before submitting.";

export type SubmissionCheck = { ok: true } | { ok: false; message: string };

export function checkSubmission(code: string): SubmissionCheck {
  if (code.trim() === "") {
    return { ok: false, message: EMPTY_SUBMISSION_MESSAGE };
  }
  return { ok: true };
}
